### PR TITLE
Integration tests

### PR DESCRIPTION
**Pull request by kderme**
ETCM-281

This pr:
- Adds new quicktest-state-machine tests, which spawn many nodes and tests properties of the system.
- Introduce a Haskell mantis mock server
- Fixes many bugs that I found while writing and running the tests
- Improves logging
- Introduces many simplifications and dead code removal.

Eventually these tests could look like https://github.com/advancedtelematic/quickcheck-state-machine/blob/master/test/RQlite.hs, so we could tests network partitions, introduce message delays etc. However this is out of scope for ETCM-281

### CODE REVIEW

**Review by NinjaTrappeur — APPROVED** [review]
Overall LGTM.

Some small nits. Also, we need to reformat everything before considering merging that.

File: morpho-checkpoint-node/src/Morpho/RPC/PoWMock.hs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import System.Timeout
+import Prelude
+
+data RPCRequst
+  = GetBlock (PoWNodeJSONRequest [Int])
+  | PutCheckpoint (PoWNodeJSONRequest PoWBlockchainCheckpoint)
+
+decodeRPCRequst :: BL.ByteString -> Either String RPCRequst
+decodeRPCRequst bs = case decode bs of
+  Just (x :: PoWNodeJSONRequest [Int]) -> Right $ GetBlock x
```
[the remark this answers is not in the document]
Comment (kderme) [reply morpho-checkpoint-node/src/Morpho/RPC/PoWMock.hs:34]: We will leave this part for later as discussed.

File: morpho-checkpoint-node/tests/Test/Morpho/QSM.hs [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+    noCleanup
+  where
+    nodes = Node <$> [0 .. (nodesNumber - 1)]
+    handlesMap = M.fromList $ zip nodes handles
+    mockMap = nodes
+
+unusedSM :: Config -> StateMachine Model Command IO Response
+unusedSM cfg = mkSM cfg $ error "NodeHandle not used on generation or shrinking"
+
+prop_1 :: Property
```
Comment (kderme) [inline morpho-checkpoint-node/tests/Test/Morpho/QSM.hs:253]: Comment what we test

**Comment by NinjaTrappeur** [comment]
Thanks!

**Pull request merged**